fedora_messaging/twisted/protocol.py

```diff
@@ -0,0 +1,296 @@
+# This file is part of fedora_messaging.
+# Copyright (C) 2018 Red Hat, Inc.
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, write to the Free Software Foundation, Inc.,
+# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+
+from __future__ import absolute_import, unicode_literals
+
+import logging
+
+import pika
+import pkg_resources
+from pika.adapters.twisted_connection import TwistedProtocolConnection
+from twisted.internet import defer, error
+# twisted.logger is available with Twisted 15+
+from twisted.python import log
+
+from .._session import get_message, get_serialized_message
+from ..exceptions import Nack, Drop, HaltConsumer, ValidationError
+
+
+_pika_version = pkg_resources.get_distribution('pika').parsed_version
+
+
+class FedoraMessagingProtocol(TwistedProtocolConnection):
+    """A Twisted Protocol for the Fedora Messaging system.
+
+    This protocol builds on the generic pika AMQP protocol to add calls
+    specific to the Fedora Messaging implementation.
+    """
+
+    name = 'FedoraMessaging:Protocol'
+
+    def __init__(self, parameters, confirms=False):
+        """Initialize the protocol.
+
+        Args:
+            parameters (pika.ConnectionParameters): The connection parameters.
+            confirms (bool): If True, all outgoing messages will require a
+                confirmation from the server, and the Deferred returned from
+                the publish call will wait for that confirmation.
+        """
+        TwistedProtocolConnection.__init__(self, parameters)
+        self._parameters = parameters
+        self._confirms = confirms
+        self._channel = None
+        self._running = False
+        self._queues = set()
+        self._message_callback = None
+        self.factory = None
+
+    @defer.inlineCallbacks
+    def connectionReady(self, res=None):
+        """Called when the AMQP connection is ready.
+        """
+        # The optional `res` argument is for compatibility with pika < 1.0.0
+        # Create channel
+        self._channel = yield self.channel()
+        log.msg("AMQP channel created", system=self.name,
+                logLevel=logging.DEBUG)
+        yield self._channel.basic_qos(prefetch_count=0, prefetch_size=0)
```

Comment:
There's a setting for QoS: https://github.com/fedora-infra/fedora-messaging/blob/fd4d4cc6a9ec75cc5af090fd81f9fc21f2d30d05/fedora_messaging/config.py#L208